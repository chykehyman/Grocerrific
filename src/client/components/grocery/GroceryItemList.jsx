import React from 'react';
import { array, func, string } from 'prop-types';

import GroceryItem from './GroceryItem';

const propTypes = {
  items: array.isRequired,
  selectedId: string.isRequired,
  action: string.isRequired,
  deleteItem: func.isRequired,
  purchaseOrDropItem: func.isRequired
};

const renderNoItemsFound = () => (
  <div className="no-items">
    <i className="fa fa-exclamation-triangle fa-3x pb-3 d-block" />
    <p className="lead">There are no items in the store</p>
  </div>
);

const renderList = (items, isProcessing, selectedId, action, deleteItem, purchaseOrDropItem) => (
  <table className="table table-hover">
    <thead>
      <tr>
        <th>&nbsp;</th>
        <th>Name</th>
        <th>Status</th>
        <th>Action</th>
      </tr>
    </thead>
    <tbody>
      {items.map(item => (
        <GroceryItem
          key={item._id}
          item={item}
          isProcessing={isProcessing}
          selectedId={selectedId}
          action={action}
          deleteItem={deleteItem}
          purchaseOrDropItem={purchaseOrDropItem} />
      ))}
    </tbody>
  </table>
);

const GroceryItemList = ({
  items, isProcessing, selectedId, action, deleteItem, purchaseOrDropItem
}) => {
  if (items.length === 0) {
    return renderNoItemsFound();
  }
  return renderList(items, isProcessing, selectedId, action, deleteItem, purchaseOrDropItem);
};

GroceryItemList.propTypes = propTypes;

export default GroceryItemList;
