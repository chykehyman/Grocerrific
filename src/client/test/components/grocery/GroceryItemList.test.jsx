import GroceryItemList from '../../../components/grocery/GroceryItemList';

describe('<GroceryItemList /> component', () => {
  const setup = () => {
    const props = {
      items: [],
      isProcessing: false,
      selectedId: 'myID123',
      action: '',
      deleteItem: jest.fn(),
      purchaseOrDropItem: jest.fn()
    };

    return {
      shallowWrapper: shallow(<GroceryItemList {...props} />),
      props
    };
  };


  it('should render without crashing', () => {
    const { shallowWrapper } = setup();

    expect(toJson(shallowWrapper)).toMatchSnapshot();
  });

  it('should show no items when items array is empty', () => {
    const { shallowWrapper } = setup();

    expect(shallowWrapper.find('p.lead').text()).toEqual('There are no items in the store');
    expect(shallowWrapper.find('GroceryItem')).toHaveLength(0);
  });

  it('should show <GroceryItem /> component when items lenght is > 0', () => {
    const { shallowWrapper, props } = setup();
    const newProps = {
      ...props,
      items: [{}, {}]
    };
    shallowWrapper.setProps(newProps);

    expect(shallowWrapper.find('p.lead')).toHaveLength(0);
    expect(shallowWrapper.find('GroceryItem')).toHaveLength(newProps.items.length);
  });
});
