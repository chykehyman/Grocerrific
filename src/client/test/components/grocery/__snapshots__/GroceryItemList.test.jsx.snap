// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<GroceryItemList /> component should render without crashing 1`] = `
<div
  className="no-items"
>
  <i
    className="fa fa-exclamation-triangle fa-3x pb-3 d-block"
  />
  <p
    className="lead"
  >
    There are no items in the store
  </p>
</div>
`;
